### ISSUE TITLE
mapped types break the generic constraints

### ISSUE THREAD

**Issue by Bigfish8** [microsoft/TypeScript#46807]
# Bug Report

### 🔎 Search Terms
generic

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ssl=27&ssc=3&pln=1&pc=1#code/C4TwDgpgBAygFgQ0gHgCoD4oF4oG8BQURUokAXFKgDSHEBuCANgK4QWr4C+++p0A8gCMAVvCQQ0mHAWJQA2gGkoASwB2UANYQQAewBmlALoUxKVIsPouPACYQAxowQAnaHuar7wZTvWvgzsp0EKYSGAAULgDmJohm6ACUZBz4do4ubh5ePn4QAUEQQqJxYeiRzjFQRaGSSZQ8APQNUPY6zq5ejCBQURCqEIH2Lb4AzgEIasAj+P6BwaHhMsR8FADkAIyrNLJNUAACUwC0EAAekF7H7W20RAwsbFDrXAmNzao6PX0DykOtqmPOCaqKZQOADCAzPJzQoiBZLIgjHQAWwgADUmKwKPDZCRwA8Nlsbji7pjHkTuJwEkA)

### 💻 Code

see playground

### 🙁 Actual behavior

no generic constraints

### 🙂 Expected behavior

correctly generic constraints

**Comment by MartinJohns** [comment]
Your code has no constraint, and nothing is broken. The type for `T` is just inferred to be `string | number`.

**Comment by Bigfish8** [comment]
Thanks for your reply, but why the type for `T` in first case is not `string | number`.

**Comment by tadhgmister** [comment]
huh, it looks like it isn't detecting the properties are co-variant in the first case, instead the generic only resolves based on the first property in the declaration. [playground](https://www.typescriptlang.org/play?#code/FAFwngDgpgBAygCwIbQIwB4AqA+GBeGAb2BlJgDMB7SgLhkwBoSyAjJAJzs2AF9RJYiFFABMWXAWJkYbTvSbSqteb2DAAJlADGAGw6xyAVwB2WkAEtKxmOygh25gG5QhacQAoOAczquoGHABKGm5NXX0KEzNLa1t7JxdkaDEcT3YfeCTRcWDuNTiHZz9Ud0IYAHpymEMAZ3NjLxhHGtQYEGQQGGQamAADJV6YTXJ6qHUKc3YakAA6OeZSJToAclRlhhgFmQ46VsqYCHZKdUMtKB6odiP2GCsYZdllmBqkMHrG806ahEpDHXGALZIEBaBB9Aa8QLAAoJPwiUoVKq1d5NGoiNodLpIHq9WSDYajcYjKazOYzLZLe5rDb7Q7HU7nGBIayXa63az9aiDF5vBowT7PH5-QHA0F9PHk6SyXaQ4BAA)

```ts
type Shape1<T> = {
    foo: T,
    bar: T
}
type Shape2<T> = {
    bar: T,
    foo: T,
}

declare function retriveShape1<T>(arg: Shape1<T>):T
declare function retriveShape2<T>(arg: Shape2<T>):T


retriveShape1({ // using vs1 that has `foo` defined first...
    foo: '1', 
    bar: 1 // produces error on 'bar' saying it should match `foo`
})
retriveShape2({ // using vs2 that has `bar` defined first....
    foo: '1', // produces an error on `foo` saying it should match `bar`.
    bar: 1
})
```

Is this a bug @MartinJohns?

**Comment by RyanCavanaugh** [comment]
This is the intended behavior, though admittedly it's not necessarily obvious why.

Generic inference *can* reject inference results that are sound. For example, if you have
```ts
declare function fn<T>(a: T, b: T): void;
```
the call
```ts
fn(0, "hello");
```
is rejected even though we *could* have inferred the type argument
```ts
fn<string | number>(0, "hello");
```
(side note: please don't use "1" and 1 as your example values, it's confusing 😅).

This is done to make generics useful, since otherwise there'd be no way to write a function that rejects the call `fn(0, "hello")`, even though this is one of the prime raison d'êtres of generics in the first place.

However, this logic has to eventually stop somewhere, since the function
```ts
declare function fn<T>(arg: T[]): void;
```
"should" be callable with a heterogeneous array:
```ts
fn([1, "two"]);
```

The question is then, given some call, are we looking at something that looks more like "two arguments with linked types", or are we looking at something that's more like "a legitimately heterogenous array" ? For an object type with simple properties, it's the former. For a mapped type, it's the latter.

Side note: there are no constraints whatsoever in this example.

**Comment by Bigfish8** [comment]
Very appreciated for your detailed reply！
